cpp/memilio/timer/basic_timer.h

```diff
@@ -34,31 +36,52 @@ class BasicTimer
     /// @brief Start the timer. Must be followed by exactly one stop.
     void start()
     {
+        should_be_running(false, "start");
+        set_running(true);
         m_start_time = get_time_now();
     }
 
     /// @brief Stop the timer and update the elapsed time. After calling stop, the timer may be started again.
     void stop()
     {
+        should_be_running(true, "stop");
+        set_running(false);
         const TimeType stop_time = get_time_now();
         m_elapsed_time += stop_time - m_start_time;
     }
 
-    /// @brief Set the elapsed time to 0. Only call while stopped.
+    /// @brief Set the elapsed time to 0. Only call while the timer is stopped.
     void reset()
     {
+        should_be_running(false, "reset");
         m_elapsed_time = mio::timing::DurationType{0};
     }
 
-    /// @brief Get the total time spent between starts and stops. Do not use while the timer is started.
+    /// @brief Get the total time spent between starts and stops. Only call while the timer is stopped.
     DurationType get_elapsed_time() const
     {
+        should_be_running(false, "get_elapsed_time");
         return m_elapsed_time;
     }
 
+    ~BasicTimer()
+    {
+        should_be_running(false, "~BasicTimer");
+    }
+
 private:
     TimeType m_start_time; ///< The last time point at which start() was called
     DurationType m_elapsed_time{0}; ///< The total time spent between starts and stops.
+
+#ifndef NDEBUG
+    bool m_is_running = false; ///< In Debug builds, tracks whether the timer is running.
```

Comment:
a public getter might be nice in debug mode

Reply:
It is usually a bad idea to have the interface depend on compiler flags, so either we add the getter and update is_running in Release as well, or we keep it inaccessible (other than through log messages).

Reply:
I would prefer keeping it inaccessible, since updating it without checks seems odd, and checking it is more expensive than I'd like.
Also, why would you need to check that the timer is running? I can only think of bad use cases. 
We can always expose the variable later, when we find a good use for it.